lua_modules/mcp23017/mcp23017.lua

```diff
@@ -0,0 +1,184 @@
+	@github https://github.com/plomi-net
+	@version 1.0.0
+]]
+
+local i2c, string, issetBit, setBit, clearBit, error = i2c, string, bit.isset, bit.set, bit.clear, error
+local isINPUT, isGPB, isHIGH = true, true, true
+
+-- registers (not used registers are commented out)
+local MCP23017_IODIRA = 0x00
+local MCP23017_IODIRB = 0x01
+local MCP23017_DEFVALA = 0x06
+local MCP23017_DEFVALB = 0x07
+local MCP23017_GPIOA = 0x12
+local MCP23017_GPIOB = 0x13
+--[[
+local MCP23017_IPOLA = 0x02
+local MCP23017_IPOLB = 0x03
+local MCP23017_GPINTENA = 0x04
+local MCP23017_GPINTENB = 0x05
+local MCP23017_DEFVALA = 0x06
+local MCP23017_DEFVALB = 0x07
+local MCP23017_INTCONA = 0x08
+local MCP23017_INTCONB = 0x09
+local MCP23017_IOCON = 0x0A
+local MCP23017_IOCON2 = 0x0B
+local MCP23017_GPPUA = 0x0C
+local MCP23017_GPPUB = 0x0D
+local MCP23017_INTFA = 0x0E
+local MCP23017_INTFB = 0x0F
+local MCP23017_INTCAPA = 0x10
+local MCP23017_INTCAPB = 0x11
+local MCP23017_OLATA = 0x14
+local MCP23017_OLATB = 0x15
+]]
+
+-- metatable
+local mcp23017 = {
+    INPUT = isINPUT,
+    OUTPUT = not isINPUT,
+    GPA = not isGPB,
+    GPB = isGPB,
+    HIGH = isHIGH,
+    LOW = not isHIGH
+}
+mcp23017.__index = mcp23017
+
+-- check device is available on address
+local function checkDevice(address, i2cId)
+    i2c.start(i2cId)
+    local response = i2c.address(i2cId, address, i2c.TRANSMITTER)
+    i2c.stop(i2cId)
+    return response
+end
+
+-- write byte
+local function writeByte(address, i2cId, registerAddr, val)
+    i2c.start(i2cId)
+    i2c.address(i2cId, address, i2c.TRANSMITTER)
+    i2c.write(i2cId, registerAddr)
+    i2c.write(i2cId, val)
+    i2c.stop(i2cId)
+end
+
+-- read byte
+local function readByte(address, i2cId, registerAddr)
+    i2c.start(i2cId)
+    i2c.address(i2cId, address, i2c.TRANSMITTER)
+    i2c.write(i2cId, registerAddr)
+    i2c.stop(i2cId)
+    i2c.start(i2cId)
+    i2c.address(i2cId, address, i2c.RECEIVER)
+    local data = i2c.read(i2cId, 1)
+    i2c.stop(i2cId)
+    return string.byte(data)
+end
+
+-- check pin is in range
+local function checkPinIsInRange(pin)
+    if pin > 7 or pin < 0 then
+        error("MCP23017 the pin must be between 0 and 7")
+    end
+    return pin
+end
+
+local function reset(address, i2cId)
+    writeByte(address, i2cId, MCP23017_IODIRA, 0xFF)
+    writeByte(address, i2cId, MCP23017_IODIRB, 0xFF)
+end
+
+-- setup device
+local function setup(address, i2cId)
+
+    -- check device address (0x20 to 0x27)
+    if (address < 32 or address > 39) then
+        error("MCP23017 address is out of range")
+    end
+
+    if (checkDevice(address, i2cId) ~= true) then
+        error("MCP23017 device on " .. string.format('0x%02X', address) .. " not found")
+    else
+        reset(address, i2cId)
+        return 1
+    end
+end
+
+return function(address, i2cId)
+    local self = setmetatable({}, mcp23017)
+
+    if setup(address, i2cId) then
+        self.writeIODIR = function(sf, bReg, newByte) -- luacheck: no unused
```

Comment:
Having these functions close over address and i2cId means that there are going to be unique closures for each mcp23017 instantiated.  Any reason not to have them on the metatable and use `self.address` and `self.i2cId`?